File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        #1. Initialize two variables (current and overall max sum)
        counterSum = 0
        overallMax = -float("inf")
        
        #2. for loop 
        for number in nums:
            counterSum += number
            
            #a. Update the overall Max
            overallMax = max(counterSum, overallMax);
            
            #b. Handle negative case of counterSum 
            if counterSum<0:
                counterSum = 0
        
        #3. return the overallMax value 
        return overallMax
        

File: 0053-maximum-subarray/0053-maximum-subarray.java
//Kadane's Algorithm - Optimal Solution
class Solution {
    public int maxSubArray(int[] nums) {
        //1. Initialize the counter variables 
        int maxCounter = Integer.MIN_VALUE;
        int sumCounter = 0;
        
        //2. For loop through nums array 
        for(int number: nums){
            //a. Increment to the sum 
            sumCounter+=number;
            
            //b. maxCounter update its value 
            maxCounter = Math.max(sumCounter, maxCounter);
            
            //c. negative value edge case 
            if (sumCounter<0){
                sumCounter = 0;
            }
        }
        
        return maxCounter;
    }
}